Producing unstable snapshots

### Describe the bug

The name of the test printed in the snapshot file may include wrongly-formed characters. On windows, it leads Vitest to consider the snapshot as obsolete and recreate a new one for each run.

**Where do the bug comes from?** -> loupe

The bug is visible in Vitest but comes from loupe and has been reported on their side too at https://github.com/chaijs/loupe/issues/78.

In Vitest the call to loupe responsible to the problem is in `@vitest/utils/dist/chunk-display.js` within `objDisplay`. Once fixed on loupe's side I can probably help and come with some dedicated tests covering that issue to avoid it to come back.

**Temporary fix**

While waiting for a fix on loupe's side, I can propose a fix on your side that will drop/replace surrogates not being in pairs? What do you think?

### Reproduction

Consider the following example:

```js
import { expect, test } from "vitest";

test.each([{ value: "\u{1f431}".repeat(50) }])(
  "print the value properly for value=$value",
  ({ value }) => {
    expect(value).toMatchSnapshot();
  }
);
```

It leads to the snapshot-ed value:

```txt
exports[`print the value properly for value='🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱�…' 1`] = `"🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱"`;
```

This snapshot name is invalid as it comes with a split-cat or more precisely a split-surrogate-pair which is not an existing character outside of JavaScript.

### System Info

```shell
System:
    OS: Windows 11 10.0.22621
    CPU: (4) x64 Intel(R) Core(TM) i5-3570K CPU @ 3.40GHz
    Memory: 1.50 GB / 7.89 GB
  Binaries:
    Node: 20.8.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.21 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 10.2.3 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Chromium (123.0.2420.97)
    Internet Explorer: 11.0.22621.1
  npmPackages:
    vitest: ^1.5.3 => 1.5.3
```


### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

The issue can easily be detected (needs possibly several retries) with the following property based test:

```js
import fc from 'fast-check'

describe('objDisplay', () => {
  test('Only produce valid strings with objDisplay', () => {
    // NOTE: The instancitation of fc.anything(...) should probably be made shorter, today it does not provide any unicode-aware
    // built-in version and thus require users to be verbose... Ideally, it should be as simple as `fc.anything({withUnicode: true})`
    fc.assert(fc.property(fc.anything({ key: fc.fullUnicodeString(), values: [
      fc.boolean(),
      fc.maxSafeInteger(),
      fc.double(),
      fc.fullUnicodeString(),
      fc.oneof(fc.fullUnicodeString(), fc.constant(null), fc.constant(undefined)),
    ] }), fc.integer({ min: 2 }), (value, truncate) => {
      expect(() => encodeURI(objDisplay(value, { truncate }))).not.toThrow()
    }))
  })
})

// FAIL  |threads| test/utils.spec.ts > objDisplay > Only produce valid strings with objDisplay
// Error: Property failed after 74 tests
// { seed: -1289013075, path: "73:1:1:2:2:2:4:3:2:2:3:2:2:2:3:2:2:2:6:5:5:7:5:6:6:6:9:7:6:7:0", endOnFailure: true }
// Counterexample: ["𐀀 ",4]
// Shrunk 30 time(s)
// Got AssertionError: expected [Function] to not throw an error but 'URIError: URI malformed' was thrown
```

If adding tests (waiting your go), I'll probably not use property based technique for now to avoid pulling something additional to your dev-deps. But I wanted to let you know that it could help to detect such issues in an automatic fashion. So I can work on adding some if you feel it might be interested (Jest found many problems thanks to fast-check).

--- a/packages/utils/package.json
+++ b/packages/utils/package.json
@@ -63,7 +63,7 @@
   "dependencies": {
     "diff-sequences": "^29.6.3",
     "estree-walker": "^3.0.3",
-    "loupe": "^3.1.0",
+    "loupe": "^3.1.1",
     "pretty-format": "^29.7.0"
   },
   "devDependencies": {

--- a/test/core/test/utils.spec.ts
+++ b/test/core/test/utils.spec.ts
@@ -1,5 +1,5 @@
 import { beforeAll, describe, expect, test } from 'vitest'
-import { assertTypes, deepClone, objectAttr, toArray } from '@vitest/utils'
+import { assertTypes, deepClone, objDisplay, objectAttr, toArray } from '@vitest/utils'
 import { deepMerge, resetModules } from '../../../packages/vitest/src/utils'
 import { deepMergeSnapshot } from '../../../packages/snapshot/src/port/utils'
 import type { EncodedSourceMap } from '../../../packages/vite-node/src/types'
@@ -269,5 +269,19 @@ describe('objectAttr', () => {
     ${{ func }}                   | ${'func'}       | ${func}
   `('objectAttr($value, $path) -> $expected', ({ value, path, expected }) => {
     expect(objectAttr(value, path)).toEqual(expected)
+  })
+})
+
+describe('objDisplay', () => {
+  test.each`
+  value | expected
+  ${'a'.repeat(100)} | ${`'${'a'.repeat(37)}…'`}
+  ${'🐱'.repeat(100)} | ${`'${'🐱'.repeat(18)}…'`}
+  ${`a${'🐱'.repeat(100)}…`} | ${`'a${'🐱'.repeat(18)}…'`}
+  `('Do not truncate strings anywhere but produce valid unicode strings for $value', ({ value, expected }) => {
+    // encodeURI can be used to detect invalid strings including invalid code-points
+    // note: our code should not split surrogate pairs, but may split graphemes
+    expect(() => encodeURI(objDisplay(value))).not.toThrow()
+    expect(objDisplay(value)).toEqual(expected)
   })
 })

